docker/compose/ibft.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2017 AMIS Technologies
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package compose
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"text/template"
+
+	"github.com/getamis/istanbul-tools/docker/service"
+)
+
+type Compose interface {
+	String() string
+}
+
+type ibft struct {
+	IPPrefix string
+	EthStats *service.EthStats
+	Services []*service.Validator
+}
+
+func New(ipPrefix string, number int, secret string, nodeKeys []string,
+	genesis string, staticNodes string, quorum bool) Compose {
+	bft := &ibft{
+		IPPrefix: ipPrefix,
+		EthStats: service.NewEthStats(fmt.Sprintf("%v.9", ipPrefix), secret),
+	}
+	bft.init(number, nodeKeys, genesis, staticNodes)
+	if quorum {
+		return newQuorum(bft, number)
+	}
+	return bft
+}
+
+func (bft *ibft) init(number int, nodeKeys []string, genesis string, staticNodes string) {
+	for i := 0; i < number; i++ {
+		s := service.NewValidator(i,
+			genesis,
+			nodeKeys[i],
+			"",
+			30303+i,
+			8545+i,
+			bft.EthStats.Host(),
+			// from subnet ip 10
+			fmt.Sprintf("%v.%v", bft.IPPrefix, i+10),
+		)
+
+		staticNodes = strings.Replace(staticNodes, "0.0.0.0", s.IP, 1)
+		bft.Services = append(bft.Services, s)
+	}
+
+	// update static nodes
+	for i := range bft.Services {
+		bft.Services[i].StaticNodes = staticNodes
+	}
+}
+
+func (bft ibft) String() string {
+	tmpl, err := template.New("ibft").Parse(ibftTemplate)
+	if err != nil {
+		fmt.Printf("Failed to parse template, %v", err)
+		return ""
+	}
+
+	result := new(bytes.Buffer)
+	err = tmpl.Execute(result, bft)
+	if err != nil {
+		fmt.Printf("Failed to render template, %v", err)
+		return ""
+	}
+
+	return result.String()
+}
+
+var ibftTemplate = `version: '3'
```

Comment:
How about making a template file instead of a string variable?

Reply:
Follow the coding style in go-ethereum. https://github.com/ethereum/go-ethereum/blob/master/accounts/abi/bind/template.go